docs/CIP_6_Pipelines_Registry.md

```diff
@@ -0,0 +1,309 @@
+
+```python
+collection = client.create_collection("my_collection", pipeline=DefaultEmbeddingFunctionPipelineStep(), output_pipeline=ReRank())
+# Add initial data
+collection.add(data=["small_chunk_1", "small_chunk_2"], ids=[0, 1])
+# Query the collection
+collection.query(data=["query"], n_results=100) # ReRank will run before the results are returned
+```
+
+NOTE: Generically, one can view the existing behavior of chroma as a pipeline that runs metadata filtering followed by vector search. In the future, we could allow exposing this basic pipeline to the user as the default.
+
+This abstraction would also allow for the full RAG loop to be implemented in Chroma. For example, if you wanted to run a query, and then generate a response using a language model, you could do something like this:
+
+```python
+collection.query(data=["query"], n_results=100, output_pipeline=Generate(Composite(prompt)))
+```
+
+This pipeline first runs a partially initialized composite step on the results of the query, and then runs the generate step on the output of the composite step. This allows for the full RAG loop to be implemented in Chroma. The composite step combines the prompt and the retrieved documents, and the generate step generates the response.
+
+#### Future work: Parallel pipeline execution
+Not all pipeline stages need to executed sequentially. In the future, we can add support for parallel pipeline execution by changing how the pipeline is defined.
+
+## Storage Pipelines
+Chroma right now implicitly stores your documents, however we also want to support multimodal embeddings and part of this CIP is to propose we deprecate documents= in favor of data=. The contract then becomes that the data you input (or that is generated by your pipeline) is stored by default and can be retrieved by the user in some form. We also want to support storing mixed modalities in a collection. For example, a CLIP based collection should be able
+to store both text and images.
+
+We propose to add Storage Pipeline steps capable of storing the inputted data into a Storage layer, and then storing a string reference to that data in the metadata of the document. This is already how the document is currently stored and backed by the metadata segment. The metadata segment stores a KV pair "chroma:document" -> the document contents. For example, say you had an image in memory and wanted to store it in Chroma, you could do something like this:
+
+```python
+collection = client.create_collection("my_collection", pipeline=DefaultEmbeddingFunctionPipelineStep(ImageStoragePipelineStep()))
+collection.add(data=[image], ids=[0]) # image is a PIL image, or numpy array etc
+
+# The ImageStoragePipelineStep would store the image in the storage layer, and then pass through the url to the image instead.
+# The metadata segment would then store the KV pair "chroma:document" -> "https://storage.com/image_1"
+```
+
+We can ship two StoragePipelines to start
```

Comment:
For some data types like images, we can also use the underlying DB - base64 the image and put it in string_value (TEXT size is 2GB for SQLite). This has implications, of course, for other RDMBS (distributed version) but so does datatypes in general.